### PR TITLE
Added support for data field in SnapdTask

### PR DESCRIPTION
**Pull request by sergio-costas**
This "data" field can contain several info, like the list of snaps that are affected by this task. Implemented in https://github.com/snapcore/snapd/pull/13953

### CODE REVIEW

**Comment by sergio-costas** [comment]
@robert-ancell Can you review this when you have some spare time, please?

**Comment by sergio-costas** [comment]
@robert-ancell Now it is ready for review. Can you check it when you have some spare time, please?

File: snapd-glib/requests/snapd-json.c [hunk 1]
```diff
@@ -763,6 +763,14 @@ _snapd_json_parse_notice (JsonNode *node, GError **error)
     return retlist;
 }
 
+static void
+add_affected_snap_to_list (JsonArray *array, guint index, JsonNode *element, void *data)
+{
+    GPtrArray *list = (GPtrArray *) data;
+    const gchar *snap_name = json_array_get_string_element (array, index);
+    g_ptr_array_add (list, g_strdup(snap_name));
```
Comment (robert-ancell) [inline snapd-glib/requests/snapd-json.c:771]: Missing whitespace. Note to self that we really should consider running clang-format over this.

File: snapd-glib/requests/snapd-json.c [hunk 2]
```diff
@@ -791,6 +799,26 @@ _snapd_json_parse_change (JsonNode *node, GError **error)
...
+            if (affected_snaps_node != NULL) {
+                if (json_node_get_value_type (affected_snaps_node) != JSON_TYPE_ARRAY) {
+                    g_set_error (error,
+                                SNAPD_ERROR,
+                                SNAPD_ERROR_READ_FAILED,
+                                "Unexpected change type");
+                    return NULL;
+                }
+                JsonArray *affected_snaps = json_node_get_array (affected_snaps_node);
+                GPtrArray *affected_snaps_array = g_ptr_array_new_full(json_array_get_length (affected_snaps), g_free);
```
Comment (robert-ancell) [inline snapd-glib/requests/snapd-json.c:815]: whitespace

File: snapd-glib/snapd-task-data.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+ *
+ * Returns: (transfer full) (allow-none) (element-type gchar*): a #GPtrArray or %NULL.
+ *
+ * Since: 1.66
+ */
+GPtrArray *
+snapd_task_data_get_affected_snaps (SnapdTaskData *self)
+{
+    g_return_val_if_fail (SNAPD_IS_TASK_DATA (self), NULL);
+    return g_ptr_array_ref(self->affected_snaps);
```
Comment (robert-ancell) [inline snapd-glib/snapd-task-data.c:60]: whitespace

File: tests/test-qt.cpp [hunk 4]
```diff
@@ -7410,11 +7410,65 @@ test_get_notices_after_notice ()
...
+    g_assert_null (data3);
+}
+
+
 int
 main (int argc, char **argv)
 {
     g_test_init (&argc, &argv, NULL);
-
+    g_test_add_func ("/task/test_task_data_field", test_task_data_field);
```
Comment (robert-ancell) [inline tests/test-qt.cpp:7471]: The empty line has been removed.

File: snapd-glib/snapd-task-data.c [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+ * snapd_task_data_get_affected_snaps:
+ * @task_data: a #SnapdTaskData.
+ *
+ * Get the list of snaps that are affected by this task, or %NULL if snapd doesn't send this data.
+ *
+ * Returns: (transfer full) (allow-none) (element-type gchar*): a #GPtrArray or %NULL.
+ *
+ * Since: 1.66
+ */
+GPtrArray *
```
Comment (robert-ancell) [inline snapd-glib/snapd-task-data.c:56]: In other cases (e.g. SnapdConnection.get_slot_attribute_names) we've returned GStrv here. I think for consistency we should do the same here.

File: snapd-glib/requests/snapd-json.c [hunk 6]
```diff
@@ -791,6 +799,26 @@ _snapd_json_parse_change (JsonNode *node, GError **error)
...
+        JsonObject *data = _snapd_json_get_object (object, "data");
+        g_autoptr(SnapdTaskData) task_data = NULL;
+        if (data != NULL) {
+            JsonNode *affected_snaps_node = json_object_get_member (data, "affected-snaps");
+            if (affected_snaps_node != NULL) {
+                if (json_node_get_value_type (affected_snaps_node) != JSON_TYPE_ARRAY) {
+                    g_set_error (error,
+                                SNAPD_ERROR,
+                                SNAPD_ERROR_READ_FAILED,
+                                "Unexpected change type");
```
Comment (robert-ancell) [inline snapd-glib/requests/snapd-json.c:811]: Copy-paste error - the text doesn't match the error.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Some small fixes and the use of GStrv over GPtrArray and this should be good to land.

File: snapd-glib/requests/snapd-json.c (hunk 2 above)
Comment (sergio-costas) [reply snapd-glib/requests/snapd-json.c:815]: Fixed.

File: snapd-glib/snapd-task-data.c (hunk 3 above)
Comment (sergio-costas) [reply snapd-glib/snapd-task-data.c:60]: Fixed.

File: tests/test-qt.cpp (hunk 4 above)
Comment (sergio-costas) [reply tests/test-qt.cpp:7471]: Fixed.

File: snapd-glib/requests/snapd-json.c (hunk 6 above)
Comment (sergio-costas) [reply snapd-glib/requests/snapd-json.c:811]: fixed.

File: snapd-glib/snapd-task-data.c (hunk 5 above)
Comment (sergio-costas) [reply snapd-glib/snapd-task-data.c]: You are right. Done.

**Comment by sergio-costas** [comment]
@robert-ancell it's ready for another round.

**Review by robert-ancell — APPROVED** [review]
LGTM

**Pull request merged**